src/data/sparse_page_source.cc

```diff
@@ -67,9 +67,10 @@ SparsePageSource::SparsePageSource(const std::string& cache_info,
     size_t fbegin = fi->Tell();
     prefetchers_[i].reset(new dmlc::ThreadedIter<SparsePage>(4));
     prefetchers_[i]->Init([&fi, &fmt] (SparsePage** dptr) {
-        if (*dptr == nullptr) {
-          *dptr = new SparsePage();
+        if (*dptr != nullptr) {
+          delete *dptr;
```

Comment:
Any particular reason for this?

Reply:
If `SparsePage`s are reused, `data` in the last page might require different sharding because the offsets are different. The current `HostDeviceVector::Reshard()` method doesn't allow this because it assumes sharding only happens once. Since we don't really care about the device side data once we are done with a `SparsePage`, it seems simpler to just delete it.

Alternatively we could rename the current `Reshard` method to `Shard`, and add a new method `Reshard` to explicitly reshard an already sharded `HostDeviceVector`. Let me know if you prefer this.

Reply:
BTW I created #4396 if we want to take the second approach.

Reply:
This approach looks much easier. There may be a performance to the current CPU code however. If the batch size is consistent the old implementation reuses memory in the SparsePage object when calling fmt->Read(). If you can do an experiment to show that the impact is negligible this approach is strongly preferred.

Reply:
No longer need this.